packages/std/src/addresses.rs

```diff
@@ -23,8 +24,18 @@ use crate::binary::Binary;
 /// This type is immutable. If you really need to mutate it (Really? Are you sure?), create
 /// a mutable copy using `let mut mutable = Addr::to_string()` and operate on that `String`
 /// instance.
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, Hash, JsonSchema)]
-pub struct Addr(String);
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, JsonSchema)]
+pub struct Addr(#[schemars(with = "String")] AddrInner);
+
+/// The maximum allowed size for bech32 (https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32)
+const ADDR_MAX_LENGTH: usize = 90;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
```

Comment:
I want to reflect on the performance changes of this. We should add some guidance on usage.

1) Passing `Addr` as an argument copies 94 bytes, rather than a 4 byte pointer (like `.clone()` of the past version, but no heap allocations).
2) Passing `&Addr` is cheap and can be restored to `Addr` via `copy()` or `clone()`.
3) `Addr`/`&Addr.as_str()` is cheap
4) `Addr.to_string()` requires a heap allocation (not just consuming the original Addr).

With this in mind, I think we can recommend returning `Addr` and passing in `&Addr` to functions.

Reply:
Yes, if we pass `Addr` references to functions (which is the recommended approach in general), and only return `Addr`, I think this would be OK, performance/memory-wise.

Reply:
I would like to play around with this a bit next week. The size of `Addr` is also my major (and only) concern here. But since we do not have huge amounts of addresses in memory at the same time and we have a well known size limit, I think it is fair to store the addresses on the stack.

We can always push them to heap by using `Box<Addr>`. Then we get what we had before, just without the mutability of a String implementation.

Reply:
What is the problem you are trying to solve?
I feel this will just make my life harder and not solve the problem that motivated AddrRef

Reply:
> What is the problem you are trying to solve?

Make using **owned** Addr easier, especially literals. Then in many places you suddenly get an Addr->&str conversion instead of the opposite direction.

I'm not claiming this solves all issues and it is well possible that it does not replace what `AddrRef` does.

Reply:
> Make using owned Addr easier, especially literals. 

I didn't have any issues there. And am afraid I will have to think way too much about unneeded copies of this struct if it happens. See #872 for another solution that provide simple literals for tests